src/main/java/roomescape/repository/ReservationDAO.java

```diff
@@ -0,0 +1,30 @@
+package roomescape.repository;
+
+import java.util.List;
+import roomescape.domain.Reservation;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class ReservationDAO {
+    private JdbcTemplate jdbcTemplate;
```

Comment:
[의견 🗣️ ]

jdbcTemplate이라는 빈을 주입받는 방식이 생성자라면 불변 선언이 가능합니다!
```suggestion
    private final JdbcTemplate jdbcTemplate;
```

Reply:
[질문 ✋ ]

의존성 주입에는 어떤 방식들이 있고, 각각의 장단점은 무엇일까요?
왜 파도는 생성자 주입을 선택하셨나요?

ref) https://ittrue.tistory.com/227

Reply:
> [의견 🗣️ ]
> 
> jdbcTemplate이라는 빈을 주입받는 방식이 생성자라면 불변 선언이 가능합니다!

아 넵.. ㅎㅎ 항상 저런 불변 선언을 계속 해야지 해야지 하는데 까먹게 되네요.. 수정하겠습니다~~

> [질문 ✋ ]
> 
> 의존성 주입에는 어떤 방식들이 있고, 각각의 장단점은 무엇일까요? 왜 파도는 생성자 주입을 선택하셨나요?
> 
> ref) https://ittrue.tistory.com/227

의존성 주입은 크게 4가지로 구성됩니다.

- **생성자 주입**
말 그대로 생성자를 통해서 의존 관계를 주입받는 방법입니다.
생성자 호출 시점에 1번만 호출되는 것을 보장할 수 있고, 불변과 필수 의존 관계에 사용되며, NPE를 방지할 수 있다는 특징이 있습니다.
참고로, 생성자가 1개만 존재하는 경우에는 @Autowired를 생략할 수 있다는 특징도 있습니다.

- **setter 주입**
setter 메서드를 통해 의존 관계를 주입받는 방법입니다.
생성자 주입과는 반대로 선택과 변경 가능성이 있는 의존 관계에 사용되고, @Autowired를 입력하지 않으면 실행되지 않습니다.

- **필드 주입**
필드에 @Autowired를 붙여서 바로 주입하는 방법입니다.
코드가 간결해진다는 장점이 있습니다. 하지만 웬만해서는 사용하지 않는데, 필드 주입을 하면 DI 컨테이너 안에서만 작동을 하게 되어 순수 자바 코드로 테스트하기 어렵고, final 키워드를 통해 불변 속성이라고 볼 수도 없고, setter로 가변 속성이라고 볼 수도 없는 애매한 상황이 발생하기 때문입니다.

- **일반 메서드 주입**
일반 메서드르 통해서 의존관계를 주입받는 방법입니다.
한 번에 여러 필드를 주입받을 수 있다는 장점이 있지만, 이 방법 또한 거의 사용하지 않습니다.
필드 주입과 비슷하게 애매한 상황이 발생할 수 있고, 여러 필드를 받을 것이라면 차라리 생성자 주입을 사용하는 것이 좋기 때문입니다.

저도 사실 이러한 의존 관계 주입방법들이 뭐가 다르고 어떤 것이 좋을지 궁금하여 찾아본 적이 있었습니다.
공부하며 내린 결론은 웬만해서는 생성자 주입 방법을 사용하고자 했습니다.
그 이유는 
- 첫 번째, 의존 관계 주입은 애플리케이션이 실행되고 종료되기 전까지 변경되지 않아야하기 때문입니다.
- 두 번째, NPE를 방지할 수 있기 때문입니다. 제가 알기로는 runtime단계의 오류가 compile단계보다 더 심각하다고 알기 때문에 에러가 발생하더라도 compile단계에서 발생하고 고치는 것이 더 낫기 때문입니다.
- 세 번째, 이것도 위의 이유와 비슷한데 순환 참조를 방지할 수 있기 때문입니다. setter 주입과 필드 주입같은 경우 실제 코드가 호출될 때까지 문제를 알 수 없지만, 생성자 주입같은 경우 문제를 알 수 있기 때문입니다.

Reply:
생성자 주입의 3가지 방법에 대해 잘 살펴보아 주셨네요! 좋습니다 파도 🌊 

그러나, 답변이 모호한 부분이 있어 이부분에 대해 몇가지 이야기와 질문을 남깁니다.

첫째, 의존관계 주입은 애플리케이션이 실행되고 종료되기 전까지 변경되지 않아야 한다? -X
- 우리가 다형성을 사용하는 상황을 생각해봅시다. 컴파일 타임에는 인터페이스 A와 의존관계가 설정이 되어있지만 실제 런타임에서는 어떻죠? A 를 구현한 인터페이스 구현체인 AImpl과 의존관계가 설정되어 있습니다. 즉 컴파일 타임 의존성과 런타임 의존성이 달라지게 됩니다.

ref) https://mangkyu.tistory.com/226

둘째, NPE를 방지할 수 있다.
-> [질문1] 왜 생성자 주입이 NPE를 방지할 수 있나요? 생성자에서 필드를 초기화하지 않으면 null로 남겨지게 되지 않나요?
->  [질문2] 필드를 private final 로 선언하면 컴파일 타임에 null 생성을 방지할 수 있지 않나요?
=> 즉, 생성자 주입이 NPE를 왜 방지할 수 있는지에 대한 이유와 런타임과 컴파일 오류와의 연관성에 대해 한번 더 설명해주시면 좋을 것 같습니다.

셋째, 순환참조를 방지할 수 있다.
-> [질문1] 순환참조라는 것이 무엇인가요?
-> [질문2] 왜 생성자 주입이 순환참조를 방지할 수 있나요?

=> 생성자 주입과 순환참조가 어떤 연관관계를 지니는지 잘 모르겠어요. 개인적으로는 생성자 주입을 사용하더라도 순환참조가 일어날 수 있다고 생각합니다.

Reply:
음... 너무 생각없이 공부하고 답변했던 것 같네요.
다시 한 번 공부하고 고쳐봤습니다!!

> 1. 의존관계 주입은 애플리케이션이 실행되고 종료되기 전까지 변경되지 않아야 한다.

알려주신 자료 잘 읽어봤습니다!  컴파일 타임 의존성과 런타임 의존성에 대해 처음 알게 되었네요.
예를 들어, 결제 방법에 대한 인터페이스를 구현한 여러 클래스가 있을 때, 컴파일 타임에는 그 인터페이스만 알고 있지만, 런타임에는 실제 구현체가 주입될 수 있겠군요. 이번에 제대로 알게 된 것 같습니다!!

> 2. NPE를 방지할 수 있다.

제가 굉장히 모호하게 답변을 했던 것 같네요. 저는 생성자 주입에서만 final을 사용하여 불변으로 만들 수 있다고 생각했습니다.
필드 주입은 객체가 생성된 후에 Spring이 값을 주입하기 때문에, final 필드에 값을 할당할 수가 없습니다. 즉, @Autowired가 동작하는 시점에는 이미 객체가 생성된 상태이므로, final을 사용할 수 없습니다.
수정자 주입도 필드가 생성된 후 setter를 통해 값이 할당되므로 final을 사용할 수 없습니다. 즉, final 필드는 한 번 초기화되면 변경할 수 없기 때문에, setter를 통한 값 변경이 불가능합니다.
그런데 생성자 주입을 사용하면, final을 사용하여 의존성이 null이 될 가능성을 차단할 수 있습니다. 그리고 객체(위에서는 JdbcTemplate의 객체)가 생성될 때  반드시 jdbcTemplate이 주입되므로, 이후에 변경될 위험이 없습니다.
그래서 결국에는 "생성자 주입 + final 필드"를 사용하게 되면 final 필드는 반드시 생성자에서 초기화되어야 하므로, 주입되지 않으면 컴파일 오류가 발생할 것입니다. 즉, 실수로 의존성을 빠뜨려도 실행 전에 문제를 발견할 수가 있습니다.

> 3. 순환참조를 방지할 수 있다.

순환참조란 A -> B -> A 형태로 두 개 이상의 빈(bean)이 서로 의존하면서 무한 루프가 발생하는 현상입니다.
예를 들어, A 객체를 생성하려면 B 객체가 필요하고 B 객체를 생성하려면 A 객체가 필요한 상황입니다.

그렇다면 왜 생성자 주입이 순환 참조를 방지할 수 있는 지에 대해서는 Spring 컨테이너의 특징 때문인 것 같습니다.
Spring 컨테이너는 모든 빈을 한 번에 생성하려고 한다고 알고 있습니다. 만약 A가 B를 필요로 하고, B가 A를 필요로 하면 생성이 불가능합니다. 즉, 애플리케이션의 실행이 불가능해집니다. 그래서 생성자 주입을 사용하면 Spring이 실행 시점에서 순환 참조를 감지하여 오류를 발생시킵니다.

반면에, 필드 주입과 setter 주입은 순환 참조가 가능합니다. @Autowired 필드 주입이나 수정자 주입을 사용하면, 실제 객체가 필요할 때 주입할 수도 있습니다. 또, 이 부분은 구글링을 하다가 알게 된 사실이긴 한데, @Lazy 애너테이션을 사용하면, Spring이 빈을 생성할 때 순환 참조를 바로 감지하지 않기 때문에 실행 중 문제가 발생할 수도 있습니다. (런타임 오류)

Reply:
[답변]
-  즉, 실수로 의존성을 빠뜨려도 실행 전에 문제를 발견할 수가 있습니다.
- > 핵심적인 부분을 잘 학습해주신 것 같습니다 😄 

> 순환참조를 발생할 수 있다.
- 생성자 주입이 순환참조를 예방해주는 것은 아닙니다. 예를 들어 A안에 B가 있고, B 안에 A는 상황을 순환참조라 정의하셨는데요.
가령 Dao가 Service를 가지고 Service가 Dao를 가진 이런 상황이 되겠죠?
```java
@Repository
public class ReservationDAO {
    private final JdbcTemplate jdbcTemplate;
    private final ReservationService reservationService;

    public ReservationDAO(JdbcTemplate jdbcTemplate, ReservationService reservationService) {
        this.jdbcTemplate = jdbcTemplate;
        this.reservationService = reservationService;
    }
```

애플리케이션을 실행시키면 파도가 말한대로 오류가 발생합니다. 순환참조가 되어 있다고 친절하게 스프링이 알려주네요!
<img width="798" alt="image" src="https://github.com/user-attachments/assets/e07f25ac-ac3b-4caa-a168-ca7c6344ae89" />

그런데 위의 Dao 로직을 보면 생성자 주입을 사용하고 있지 않나요? 순환참조로 인해 에러가 발생하는 것을 생성자 주입이 예방해준다는 논리가 성립하려면, 생성자 주입을 사용한다 -> 순환참조가 발생할 일이 없다로 이어져야 하는데, 해당 상황에서는 생성자 주입을 사용하고 있음에도 순환참조가 발생하고 있습니다. 여기에 대해 "에러 발생했잖아요?!" 라고 말씀하실 수도 있는데요. 해당 에러는 순환참조가 발생하여 빈을 띄우지 못하는 상황에 대한 에러이지 의존 주입이 생성자인 것과 크게 관계가 없습니다.

 
다만 해당 내용은 현재 미션의 주요 파트가 아니기 때문에 자세한 내용은 다음 미션에서 더 이야기나눠보도록 하죠!